### PR TITLE
main: add support for stlink-dap programmer

### PR DESCRIPTION
**Pull request by sago35**
STLINK could not be used if swd was specified for `openocd-transport`.

```
$ tinygo flash --target wioterminal --size short --programmer stlink-dap -x ./b1
openocd -f interface/stlink-dap.cfg -c 'transport select swd' -f target/atsame5x.cfg -c 'program C:/Users/sago3/AppData/Local/Temp/tinygo3166121016/main.hex verify reset exit'
Open On-Chip Debugger 0.11.0-rc2+dev-gcafa0b5 (2021-02-20-22:09)
Error: Debug adapter doesn't support 'swd' transport
```

`-ocd-commands` allows additional configuration, but the existence of `transport select swd` will result in an error.

```
$ tinygo flash --target wioterminal --size short --programmer stlink-dap --ocd-commands "transport select dapdirect_swd" -x ./b1
openocd -f interface/stlink-dap.cfg -c 'transport select dapdirect_swd' -c 'transport select swd' -f target/atsame5x.cfg -c 'program C:/Users/sago3/AppData/Local/Temp/tinygo2520453502/main.hex verify reset exit'
Open On-Chip Debugger 0.11.0-rc2+dev-gcafa0b5 (2021-02-20-22:09)
dapdirect_swd
Error: Can't change session's transport after the initial selection was made
```

So, in this PR, it is first necessary to be able to override `openocd-transport` with `dapdirect_swd`, etc.
Previously, only `empty-string` or `swd` could be specified.

However, openocd allows `hla_swd` and `dapdirect_swd` to be specified, as shown in the following page.
https://openocd.org/doc/html/Debug-Adapter-Configuration.html

```
# ./wioterminal-stlink.json
{
    "inherits": ["wioterminal"],
    "openocd-transport": "dapdirect_swd"
}
```

```
$ tinygo flash --target ./wioterminal-stlink.json --size short --programmer stlink-dap -x ./b1
openocd -f interface/stlink-dap.cfg -c 'transport select dapdirect_swd' -f target/atsame5x.cfg -c 'program C:/Users/sago3/AppData/Local/Temp/tinygo4040189479/main.hex verify reset exit'
Open On-Chip Debugger 0.11.0-rc2+dev-gcafa0b5 (2021-02-20-22:09)
...
(OK)
```

This PR will allow the following cheap STLINK-V2 to be used with samd21, samd51, nrf, stm32f4 and others.
image

### CODE REVIEW

**Comment by sago35** [comment]
Note that `openocd-transport` does not seem to need to be configured except when using `--programmer jlink`.
It is also possible to use `-ocd-commands 'transport select swd'` to set up jlink when using jlink.

Perhaps a configuration value like `openocd-jlink-transport` would make everyone happy, I think.

**Comment by aykevl** [comment]
While this PR seems fine, I think there is a better solution. Namely by using `dapdirect_swd` automatically when the interface is ST-Link. Untested example:

```diff
                args = append(args, "-c", cmd)
        }
        if c.Target.OpenOCDTransport != "" {
-               args = append(args, "-c", "transport select "+c.Target.OpenOCDTransport)
+               transport := c.Target.OpenOCDTransport
+               switch c.Target.OpenOCDInterface {
+               case "stlink", "stlink-v2", "stlink-v2-1":
+                       transport = "dapdirect_swd"
+               }
+               args = append(args, "-c", "transport select "+transport)
        }
        args = append(args, "-f", "target/"+c.Target.OpenOCDTarget+".cfg")
        return args, nil
```

This change means that you don't have to set the transport manually.

File: compileopts/config.go [hunk 1]
```diff
@@ -457,7 +457,14 @@ func (c *Config) OpenOCDConfiguration() (args []string, err error) {
 		args = append(args, "-c", cmd)
 	}
 	if c.Target.OpenOCDTransport != "" {
-		args = append(args, "-c", "transport select "+c.Target.OpenOCDTransport)
+		transport := c.Target.OpenOCDTransport
+		if transport == "swd" {
+			switch openocdInterface {
+			case "stlink-dap":
```
Comment (sago35) [inline compileopts/config.go:463]: `interface/stlink.cfg` did not support `dapdirect_swd`.
Therefore, the only branch of the source code was stlink-dap.

```
$ tinygo flash --target wioterminal --size short --programmer stlink examples/blinky1
Error: Debug adapter doesn't support 'dapdirect_swd' transport

$ tinygo flash --target wioterminal --size short --programmer stlink-v2 examples/blinky1
Error: Debug adapter doesn't support 'dapdirect_swd' transport

$ tinygo flash --target wioterminal --size short --programmer stlink-v2-1 examples/blinky1
Error: Debug adapter doesn't support 'dapdirect_swd' transport
```

File: compileopts/config.go [hunk 2]
```diff
@@ -457,7 +457,14 @@ func (c *Config) OpenOCDConfiguration() (args []string, err error) {
 		args = append(args, "-c", cmd)
 	}
 	if c.Target.OpenOCDTransport != "" {
-		args = append(args, "-c", "transport select "+c.Target.OpenOCDTransport)
+		transport := c.Target.OpenOCDTransport
+		if transport == "swd" {
```
Comment (sago35) [inline compileopts/config.go:461]: In the future, when a transport other than swd is added, a bug will occur that is hard to detect, so we will check if it is swd or not.

**Comment by sago35** [comment]
@aykevl 
Ready for review.

This PR change affects only the following

* `-programmer stlink-dap`
* When trasnport is swd.


I confirmed the behavior with the following commands.

```
$ tinygo flash --target wioterminal --programmer stlink-dap examples/blinky1
Open On-Chip Debugger 0.11.0-rc2+dev-gcafa0b5 (2021-02-20-22:09)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
dapdirect_swd
Info : STLINK V2J40S7 (API v2) VID:PID 0483:3748
Info : Target voltage: 3.212698
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : clock speed 1800 kHz
Info : stlink_dap_op_connect(connect)
Info : SWD DPIDR 0x2ba01477
Info : atsame5.cpu: hardware has 6 breakpoints, 4 watchpoints
Info : starting gdb server for atsame5.cpu on 3333
Info : Listening on port 3333 for gdb connections
target halted due to debug-request, current mode: Thread
xPSR: 0x01000000 pc: 0x0000056c msp: 0x2000d6a0
** Programming Started **
Info : SAM MCU: SAMD51P19A (512KB Flash, 192KB RAM)
** Programming Finished **
** Verify Started **
** Verified OK **
** Resetting Target **
shutdown command invoked
```

File: compileopts/config.go (hunk 1 above)
Comment (aykevl) [reply compileopts/config.go:463]: Ah I see, I did not read your original message correctly.

File: compileopts/config.go (hunk 2 above)
Comment (aykevl) [reply compileopts/config.go:461]: Thanks, that's a good idea.

**Comment by aykevl** [comment]
Thank you for this improvement!

**Pull request merged**